keps/0004-add-testing-infrastructure.md

```diff
@@ -68,21 +80,48 @@ This is a two phase proposal.
 
 ### Tests
 
-Add testing packages to:
+KUDO will be tested by several different test suites that validate individual components, KUDO, and frameworks.
+
+#### Unit tests
+
+Unit tests test a single component and require no network connectivity. These should be fast, run on every pull request,
+and test that individual KUDO components behave correctly. Unit tests are written as standard Go tests.
+
+All packages should incorporate unit tests that validate their behavior. Some examples of tests that should be included are:
 
 - pkg/apis - Conformance of objects, serialization, validation
 - pkg/controller - Controllers implement the correct process flows for changes in Instance objects, and creation of PlanExecutions
 - pkg/util - Util functions behave as expected
 
-Integration tests
+Go coverage reports will be used to ensure that changes do not reduce test coverage.
+
+#### Integration tests
+
+Integration tests test the KUDO controller in its entirety. These tests can be run against either a real Kubernetes cluster
+or a local control plane. Initially, all integration tests will be run against every pull request. If this begins to degrade
+velocity, integration tests can be run on master or nightly builds. Integration tests can be written either in Go or using
+the test harness designed in KEP-0008.
+
+The integration tests will consist of a set of representative Frameworks and FrameworkVersions that utilize real world
+KUDO features.
 
-- Identify and build a common set of Frameworks and FrameworkVersions that can demonstrate/validate that functionality of the controller
-  exists in real world situations.
+Integration tests will be hidden behind a Go build tag and will only run when the `integration` tag is specified. Because
+build tags are used to separate integration from unit tests, integration tests written in Go must not be in the same file
+as unit tests.
+
+#### Framework tests
```

Comment:
What version of KUDO is used to test frameworks?  The most recent release, master?  How do we release the binary into the Framework test suite?

Reply:
There's (at least) two ways to do this:

* The test harness includes a built version of KUDO that can be started for the tests.
* Use an already deployed KUDO in the cluster.

I was envisioning that our test runs in CI would use the KUDO included in the test harness and frameworks would be tested using the latest version of the test harness. But, if we don't like that method for some reason (perhaps we want to test our manifests and built images), then we can instead deploy the latest released version of KUDO using traditional manifests prior to running the tests.

Reply:
shouldn't we always test frameworks against latest stable release of KUDO? And then before release re-run tests with the version we're planning to release to make sure we did not break anything.

Reply:
I mean if we're talking about tests that will be run when someone submits a PR against frameworks, that makes sense to me